floors/fetcher.go

```
@@ -0,0 +1,294 @@
+			}
+		case <-f.done:
+			glog.Info("Price Floor fetcher terminated")
+			return
+		}
+	}
+}
+
+func fetchAndValidate(config config.AccountFloorFetch) (*openrtb_ext.PriceFloorRules, int) {
+	floorResp, maxAge, err := fetchFloorRulesFromURL(config)
+	if err != nil {
+		glog.Errorf("Error while fetching floor data from URL: %s, reason : %s", config.URL, err.Error())
+		return nil, 0
+	}
+
+	if len(floorResp) > (config.MaxFileSize * 1024) {
+		glog.Errorf("Recieved invalid floor data from URL: %s, reason : floor file size is greater than MaxFileSize", config.URL)
+		return nil, 0
+	}
+
+	var priceFloors openrtb_ext.PriceFloorRules
+	if err = json.Unmarshal(floorResp, &priceFloors.Data); err != nil {
+		glog.Errorf("Recieved invalid price floor json from URL: %s", config.URL)
+		return nil, 0
+	} else {
+		err := validateRules(config, &priceFloors)
+		if err != nil {
+			glog.Errorf("Validation failed for floor JSON from URL: %s, reason: %s", config.URL, err.Error())
+			return nil, 0
+		}
+	}
+
+	return &priceFloors, maxAge
+}
+
+// fetchFloorRulesFromURL returns a price floor JSON and time for which this JSON is valid
+// from provided URL with timeout constraints
+func fetchFloorRulesFromURL(configs config.AccountFloorFetch) ([]byte, int, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(configs.Timeout)*time.Millisecond)
+	defer cancel()
+
+	httpReq, err := http.NewRequestWithContext(ctx, http.MethodGet, configs.URL, nil)
+	if err != nil {
+		return nil, 0, errors.New("error while forming http fetch request : " + err.Error())
+	}
+
+	httpResp, err := http.DefaultClient.Do(httpReq)
+	if err != nil {
+		return nil, 0, errors.New("error while getting response from url : " + err.Error())
+	}
+
+	if httpResp.StatusCode != 200 {
+		return nil, 0, errors.New("no response from server")
+	}
+
+	var maxAge int
+	if maxAgeStr := httpResp.Header.Get("max-age"); maxAgeStr != "" {
+		maxAge, _ = strconv.Atoi(maxAgeStr)
```

Please do not ignore errors. The http header could be malformed.

---

Done